src/analysis/processing/qgsalgorithmpointstopaths.cpp

```diff
@@ -0,0 +1,334 @@
+  if ( groupExpression.hasParserError() )
+    throw QgsProcessingException( groupExpression.parserErrorString() );
+
+  QgsFields outputFields = QgsFields();
+  if ( ! groupExpressionString.isEmpty() )
+  {
+    requiredFields.append( groupExpression.referencedColumns().values() );
+    const QgsField field = groupExpression.isField() ? source->fields().field( requiredFields.last() ) : QString( "group" );
+    outputFields.append( field );
+  }
+  outputFields.append( QgsField( "begin", orderFieldType ) );
+  outputFields.append( QgsField( "end", orderFieldType ) );
+
+  const bool naturalSort = parameterAsBool( parameters, QStringLiteral( "NATURAL_SORT" ), context );
+  QCollator collator;
+  collator.setNumericMode( true );
+
+  QgsWkbTypes::Type wkbType = QgsWkbTypes::LineString;
+  if ( QgsWkbTypes::hasM( source->wkbType() ) )
+    wkbType = QgsWkbTypes::addM( wkbType );
+  if ( QgsWkbTypes::hasZ( source->wkbType() ) )
+    wkbType = QgsWkbTypes::addZ( wkbType );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, outputFields, wkbType, source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  const QString textDir = parameterAsString( parameters, QStringLiteral( "OUTPUT_TEXT_DIR" ), context );
+  if ( ! textDir.isEmpty() &&
+       ! QDir( textDir ).exists() )
+    throw QgsProcessingException( QObject::tr( "The text output directory does not exist" ) );
+
+  // Store the points in a hash with the group identifier as the key
+  QHash< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > allPoints;
+
+  QgsFeatureRequest request = QgsFeatureRequest().setSubsetOfAttributes( requiredFields, source->fields() );
+  QgsFeatureIterator fit = source->getFeatures( request, QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+  QgsFeature f;
+  const double totalPoints = source->featureCount() > 0 ? 100.0 / source->featureCount() : 0;
+  long currentPoint = 0;
+  feedback->setProgressText( QObject::tr( "Loading points…" ) );
+  while ( fit.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( currentPoint * totalPoints );
+
+    if ( f.hasGeometry() && ! f.geometry().isNull() )
+    {
+      expressionContext.setFeature( f );
+      const QVariant orderValue = orderExpression.evaluate( &expressionContext );
+      const QVariant groupValue = groupExpressionString.isEmpty() ? QVariant() : groupExpression.evaluate( &expressionContext );
+
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
+      const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet()->clone() );
```

Comment:
```suggestion
      QgsPoint point( *qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet() ) ) );
```

But warning -- this will crash if a multipoint source is used, so we should handle the case here that the feature's geometry is multipoint and add all those points if so

Reply:
ok, I also switched to `std::stable_sort` so the point order within a multipoint is preserved; we have no way to manipulate it since the points share attributes.